### PR TITLE
[Eljer Chua] iP

### PR DESCRIPTION
**Pull request by asdzxc270697**

### CODE REVIEW

File: src/main/java/Command.java [hunk 1]
```diff
@@ -0,0 +1,3 @@
+public enum Command {
+    LIST, BYE, DONE, TODO, DEADLINE, EVENT, NULL
+}
```
Comment (Yuxinn-J) [inline src/main/java/Command.java]: I like the enum type you use here. It is very clear and intuitive to me.

File: src/main/java/Parser.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        boolean isTodo = command.equals("todo");
+        if(isTodo) return Command.TODO;
+
+        boolean isDeadline = command.equals("deadline");
+        if(isDeadline) return Command.DEADLINE;
+
+        boolean isEvent = command.equals("event");
+        if(isEvent) return Command.EVENT;
+
```
Comment (Yuxinn-J) [inline src/main/java/Parser.java]: Is it better to use switch statement here?
```suggestion
        switch(command){
        case "list":
              return Command.LIST;
        case "bye":
             return Command.BYE;
             ....
        }

```

File: src/main/java/Parser.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        return Command.NULL;
+    }
+
+    public static String processTodo(String userInput) {
+        String description = null;
+        try {
+            String[] words = userInput.split(" ", 2);
```
Comment (Yuxinn-J) [inline src/main/java/Parser.java]: Wow a clever method to deal with String, I learnt it from you!

**Review by Yuxinn-J — COMMENTED** [review]
LGTM, almost. Just give some personal advice

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,48 @@
...
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        printWelcomeMessage();
+        TaskManager taskManager = new TaskManager();
+
+        boolean isExit = false;
+        Scanner in = new Scanner(System.in);
+        do {
+            String userInput = in.nextLine();
+            String[] words = userInput.split(" ");
```
Comment (Yuxinn-J) [inline src/main/java/Duke.java]: It is better use another way to extract the command, since you create a array of String, but only use the first element.
```suggestion
            String word = userInput.substring(0, userInput.indexOf(" "););
```

File: src/main/java/Command.java [hunk 5]
```diff
@@ -0,0 +1,3 @@
+public enum Command {
+    LIST, BYE, DONE, TODO, DEADLINE, EVENT
```
Comment (ruyian) [inline src/main/java/Command.java]: Good job in capitalizing all the enums.

File: src/main/java/Event.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+public class Event extends Todo {
+    protected String at;
```
Comment (ruyian) [inline src/main/java/Event.java]: The naming "at" might be a little be confusing. It might be better if you can mention something like eventTime.

File: src/main/java/Parser.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        boolean isDone = command.equals("done");
+        if(isDone) return Command.DONE;
+
+        boolean isDeadline = command.equals("deadline");
+        if(isDeadline) return Command.DEADLINE;
+
+        boolean isEvent = command.equals("event");
+        if(isEvent) return Command.EVENT;
+
+        return Command.TODO;
```
Comment (ruyian) [inline src/main/java/Parser.java]: For readability, it is better to put all the locally declared boolean variables together and then the if statement. I find it a bit confusing to look at the first glance.

File: src/main/java/TaskManager.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        System.out.println("Nice! I've marked this task as done:\n [X] " + tasks[index].getDescription());
+    }
+
+    public void addTask(String task) {
+        tasks[taskCount + 1] = new Task(task);
+        taskCount++;
+        System.out.println("Added task to list: " + task);
+    }
+
+}
```
Comment (ruyian) [inline src/main/java/TaskManager.java]: Great readability and naming! All the methods and fields are clearly named and I can tell what you are trying to express in this class.

**Review by ruyian — COMMENTED** [review]
Great job, Eljer!

File: src/main/java/Todo.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+public class Todo extends Task {
```
Comment (kum-wh) [inline src/main/java/Todo.java]: Class names should be in PascalCase. ToDo

File: src/main/java/Task.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+    public void setDone() {
+        isDone = true;
+    }
+
+    @Override
+    public String toString() {
+        if (isDone) {
+            return "[X] " + description;
+        } else {
```
Comment (kum-wh) [inline src/main/java/Task.java]: Not necessary to use if-else on return statements.

File: src/main/java/Parser.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
+public class Parser {
+    public static Command processCommand(String command) {
+        boolean isList = command.equals("list");
+        if(isList) return Command.LIST;
+
+        boolean isBye = command.equals("bye");
+        if(isBye) return Command.BYE;
```
Comment (kum-wh) [inline src/main/java/Parser.java]: Conditional should be put on a separate line.

File: src/main/java/Parser.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
+public class Parser {
+    public static Command processCommand(String command) {
+        boolean isList = command.equals("list");
+        if(isList) return Command.LIST;
```
Comment (kum-wh) [inline src/main/java/Parser.java]: Single statement conditionals should be wrapped by curly brackets.

File: src/main/java/Duke.java [hunk 13]
```diff
@@ -1,10 +1,48 @@
...
+                taskManager.addTodoTask(userInput);
+                break;
+            default:
+                System.out.println("Sorry I don't understand the command: " + userInput);
+            }
+            System.out.println(DIVIDER);
+        } while (!isExit);
+    }
+
+    private static void printWelcomeMessage() {
```
Comment (kum-wh) [inline src/main/java/Duke.java]: Good choice of naming for the method, making it easy to understand the code

**Review by kum-wh — COMMENTED** [review]
Java coding standard is mostly followed. Code is easy to understand in general.

File: src/main/java/duke/Duke.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Scanner;
+
+public class Duke {
+    public static final String FILE_PATH = "duke.txt";
+    public static final String DIVIDER = "-----------------------------------------";
+
+    public static void main(String[] args) {
```
Comment (stephlewyh) [inline src/main/java/duke/Duke.java]: While you have abstracted out the file loading and welcome message into separate methods, the `main()` block is still too long ([>30 loC](https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods)). 

Hint: You could attempt to [ensure SLAP ](https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#slap-hard) to keep your main method short and easy to read.

File: src/main/java/duke/Duke.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+                case TODO:
+                    taskManager.addToDoTask(userInput);
+                    taskManager.printAddSuccess();
+                    break;
+                case DELETE:
+                    taskManager.deleteTask(userInput);
+                    break;
+                default:
+                    throw new DukeException("I'm sorry, but I don't know what that means :-(");
+                }
```
Comment (stephlewyh) [inline src/main/java/duke/Duke.java]: The `switch-case` layout is well done here but it takes up more lines in the main method which has other things going on.

File: src/main/java/duke/Duke.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+            } finally {
+                updateSavedFile(taskManager);
+                System.out.println(DIVIDER);
+            }
+        } while (!isExit);
+    }
+
+    private static TaskManager loadSavedFile() {
+        try {
+            return Storage.readFile(FILE_PATH);
```
Comment (stephlewyh) [inline src/main/java/duke/Duke.java]: It would be better to assign a name to the file that is being returned, instead of putting the method call at the return statement.

It is slightly confusing as to whether a TaskManger object or an actual file is being returned.

File: src/main/java/duke/parser/Parser.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        return information;
+    }
+
+    /**
+     * @param userInput
+     * @return
+     * description = information[0]
+     * <p>
+     * event time = information[1]
```
Comment (stephlewyh) [inline src/main/java/duke/parser/Parser.java]: This deviates from the Java standard in that it does not give a clear description of what the method does.  

Refer to [javadocs examples](https://se-education.org/guides/conventions/java/basic.html#comments), which gives what is returned and the input parameters in plain English sentences.

File: src/main/java/duke/parser/Parser.java [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+            int atIndex = words[1].indexOf("/at");
+            if (atIndex == -1) {
+                throw new DukeException("The event time of the event task cannot be empty.");
+            }
+
+            information[0] = words[1].substring(0, atIndex).trim();
+            information[1] = words[1].substring(atIndex + 3).trim();
+            if (information[0].equals("") || information[1].equals("")) {
+                throw new DukeException("The description or dateline time cannot be empty.");
+            }
```
Comment (stephlewyh) [inline src/main/java/duke/parser/Parser.java]: Using conditions that check for word length and index number does not [directly tell how the event will be processed](https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#structure-code-logically).  

Perhaps you could [name](https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#guideline-name-well) the boolean expressions with something more meaningful to indicate what are the valid inputs needed to process an event.

File: src/main/java/duke/Duke.java [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.Locale;
+import java.util.Scanner;
+
+public class Duke {
+    public static final String FILE_PATH = "duke.txt";
+    public static final String DIVIDER = "-----------------------------------------";
+
+    public static void main(String[] args) {
+        printWelcomeMessage();
+        TaskManager taskManager = loadSavedFile();
```
Comment (stephlewyh) [inline src/main/java/duke/Duke.java]: You wish to initialize a taskManager but are loading a saved file instead. Consider a replacing `loadSavedFile()` with a more [meaningful name](https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#use-nouns-for-things-and-verbs-for-actions`. ).

Hint: What inputs are required to construct a taskManager?

File: src/main/java/duke/parser/Parser.java [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+            information[0] = words[1].substring(0, byIndex).trim();
+            information[1] = words[1].substring(byIndex + 3).trim();
+            if (information[0].equals("") || information[1].equals("")) {
+                throw new DukeException("The description or dateline time cannot be empty.");
+            }
+        } catch (DukeException e) {
+            System.out.println(e.getMessage());
+        }
+        return information;
```
Comment (stephlewyh) [inline src/main/java/duke/parser/Parser.java]: Several input conditions here could be [named better](https://nus-cs2113-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#basic-3) to increase code readability. Review the `if` conditions you  have set up and consider assigning a name for each.

File: src/main/java/duke/task/TaskManager.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
+package duke.task;
+
+import duke.parser.Parser;
+
+import java.util.ArrayList;
+
+public class TaskManager {
```
Comment (stephlewyh) [inline src/main/java/duke/task/TaskManager.java]: As the TaskManager is performing a number of important functions, it would help to provide more [method headers](https://se-education.org/guides/conventions/java/basic.html#comments) here to describe what each one if doing.
